src/async.c

```diff
@@ -197,12 +189,20 @@ mpd_async_read(struct mpd_async *async)
 
 	if (!mpd_buffer_make_room(&async->input, min_size)) {
 		mpd_error_code(&async->error, MPD_ERROR_OOM);
+		mpd_error_message(&async->error,
+				  "Out of memory for input buffer");
 		return false;
 	}
 	room = mpd_buffer_room(&async->input);
 
-	nbytes = recv(async->fd, mpd_buffer_write(&async->input), room,
-		      MSG_DONTWAIT);
+	write_p = mpd_buffer_write(&async->input);
```

Comment:
Why can `mpd_buffer_write()` fail, after `mpd_buffer_room()` has already verified that buffer space is available?
I'm puzzled.

Reply:
Because `mpd_buffer_room()` does not initialize the data, that is the responsibility of `mpd_buffer_write()`

Reply:
This design was your idea. And I say it's overly complicated, complex and thus error prone.